File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No69SqrtX.java
package com.hazza.LeetCodeByJava.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/sqrtx/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-23
 * Time: 9:59 AM
 */
public class No69SqrtX {
    //solution 1
    public int mySqrt(int x) {
        int left = 1, right = x;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (middle == x / middle) return middle;
            else if (middle < x / middle) left = middle + 1;
            else right = middle - 1;
        }

        return right;
    }

    // solution 2
    public int mySqrt2(int x) {
        double r = x;
        while (Math.abs(r * r - x) > 0.0001)
            r = (r + x / r) / 2;
        return (int) r;
    }
}

File: src/main/python/com/hazza/leetcode_by_python/SegmentTwo/No69_Sqrt_x.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/21 20:27
# https://leetcode.com/problems/sqrtx/


class Solution:
    def mySqrt1(self, x: int) -> int:
        l, r = 1, x

        while l <= r:
            m = l + (r - l) // 2
            product = m * m
            if product == x:
                return m
            elif product > x:
                r = m - 1
            else:
                l = m + 1

        return r

    def mySqrt2(self, x: int) -> int:
        r = x

        while abs(r * r - x) > 0.0001:
            r = (r + x / r) / 2

        return int(r)


if __name__ == '__main__':
    solution = Solution()

    assert solution.mySqrt1(4) == 2
    assert solution.mySqrt1(8) == 2

    assert solution.mySqrt2(4) == 2
    assert solution.mySqrt2(8) == 2
